crates/storage/src/collections/vector.rs

```diff
@@ -0,0 +1,325 @@
+use core::borrow::Borrow;
+use core::marker::PhantomData;
+
+use borsh::{BorshDeserialize, BorshSerialize};
+
+// fixme! macro expects `calimero_storage` to be in deps
+use crate::address::{Id, Path};
+use crate::collections::error::StoreError;
+use crate::entities::{Data, Element};
+use crate::interface::{Interface, StorageError};
+use crate::{self as calimero_storage, AtomicUnit, Collection};
+
+/// A vector collection that stores key-value pairs.
+#[derive(AtomicUnit, Clone, Debug, Eq, PartialEq, PartialOrd)]
+#[type_id(251)]
+#[root]
+pub struct Vector<V> {
+    /// The entries in the vector.
+    entries: Entries<V>,
+    /// The storage element for the vector.
+    #[storage]
+    storage: Element,
+}
+
+/// A collection of entries in a vector.
+#[derive(Collection, Clone, Debug, Eq, PartialEq, PartialOrd)]
+#[children(Entry<V>)]
+struct Entries<V> {
+    /// Helper to associate the generic types with the collection.
+    _priv: PhantomData<V>,
+}
+
+/// An entry in a vector.
+#[derive(AtomicUnit, Clone, Debug, Eq, PartialEq, PartialOrd)]
+#[type_id(250)]
+pub struct Entry<V> {
+    /// The value for the entry.
+    value: V,
+    /// The storage element for the entry.
+    #[storage]
+    storage: Element,
+}
+
+impl<V: BorshSerialize + BorshDeserialize> Vector<V> {
+    /// Create a new vector collection.
+    ///
+    /// # Errors
+    ///
+    /// If an error occurs when interacting with the storage system, or a child
+    /// [`Element`](crate::entities::Element) cannot be found, an error will be
+    /// returned.
+    ///
+    pub fn new() -> Result<Self, StoreError> {
+        let id = Id::random();
+        let mut this = Self {
+            entries: Entries::default(),
+            storage: Element::new(
+                &Path::new(format!("::unused::vector::{id}::path"))?,
+                Some(id),
+            ),
+        };
+
+        let _ = Interface::save(&mut this)?;
+
+        Ok(this)
+    }
+
+    /// Add a value to the end of the vector.
+    ///
+    /// # Errors
+    ///
+    /// If an error occurs when interacting with the storage system, or a child
+    /// [`Element`](crate::entities::Element) cannot be found, an error will be
+    /// returned.
+    ///
+    pub fn push(&mut self, value: V) -> Result<bool, StoreError> {
```

Comment:
do we need to return a `bool` here?

Reply:
Implemented to be same as rust std collection

Reply:
https://doc.rust-lang.org/std/vec/struct.Vec.html#method.push